graphbolt/src/cuda/extension/gpu_graph_cache.cu

```diff
@@ -0,0 +1,320 @@
+    const int64_t num_edges, const int64_t threshold,
+    torch::ScalarType indptr_dtype, std::vector<torch::ScalarType> dtypes) {
+  const int64_t num_nodes = 1024;
+  AT_DISPATCH_INDEX_TYPES(
+      dtypes.at(0), "GpuGraphCache::GpuGraphCache", ([&] {
+        auto map_temp = map_t<index_t>{
+            num_nodes,
+            load_factor,
+            cuco::empty_key{static_cast<index_t>(-1)},
+            cuco::empty_value{std::numeric_limits<index_t>::lowest()},
+            {},
+            probing_t<index_t>{},
+            {},
+            {},
+            allocator_t<index_t>{},
+            cuco::cuda_stream_ref{cuda::GetCurrentStream()}};
+        map_ = new map_t<index_t>{std::move(map_temp)};
+      }));
+  C10_CUDA_KERNEL_LAUNCH_CHECK();  // Check the map constructor's success.
+  const auto options = torch::TensorOptions().device(c10::DeviceType::CUDA);
+  threshold_ = threshold;
+  device_id_ = cuda::GetCurrentStream().device_index();
+  map_size_ = 0;
+  num_nodes_ = 0;
+  num_edges_ = 0;
+  indptr_ = torch::zeros(num_nodes + 1, options.dtype(indptr_dtype));
+  for (auto dtype : dtypes) {
+    cached_edge_tensors_.push_back(
+        torch::empty(num_edges, options.dtype(dtype)));
+  }
+}
+
+GpuGraphCache::~GpuGraphCache() {
+  AT_DISPATCH_INDEX_TYPES(
+      cached_edge_tensors_.at(0).scalar_type(), "GpuGraphCache::GpuGraphCache",
+      ([&] { delete reinterpret_cast<map_t<index_t>*>(map_); }));
+}
+
+struct AddTuple {
+  template <typename T>
+  __host__ __device__ auto operator()(const T& a, const T& b) const {
+    return thrust::make_tuple(
+        thrust::get<0>(a) + thrust::get<0>(b),
+        thrust::get<1>(a) + thrust::get<1>(b));
+  }
+};
+
+std::tuple<
+    torch::Tensor, std::vector<torch::Tensor>, torch::Tensor, torch::Tensor,
+    torch::Tensor, int64_t>
+GpuGraphCache::Query(torch::Tensor seeds) {
+  auto allocator = cuda::GetAllocator();
+  auto index_dtype = cached_edge_tensors_.at(0).scalar_type();
+  const dim3 block(BLOCK_SIZE);
+  const dim3 grid((seeds.size(0) + BLOCK_SIZE - 1) / BLOCK_SIZE);
+  return AT_DISPATCH_INDEX_TYPES(
+      index_dtype, "GpuGraphCache::Query", ([&] {
+        auto map = reinterpret_cast<map_t<index_t>*>(map_);
+        while ((map_size_ + seeds.size(0) >= map->capacity() * load_factor)) {
+          map->rehash_async(
+              map->capacity() * growth_factor,
+              cuco::cuda_stream_ref{cuda::GetCurrentStream()});
+        }
+        auto positions =
+            torch::empty(seeds.size(0), seeds.options().dtype(index_dtype));
+        CUDA_KERNEL_CALL(
+            _QueryAndIncrement, grid, block, 0,
+            static_cast<int64_t>(seeds.size(0)),
+            static_cast<index_t>(threshold_), seeds.data_ptr<index_t>(),
+            positions.data_ptr<index_t>(), map->ref(cuco::insert_and_find));
+        auto num_cache_enter =
+            allocator.AllocateStorage<thrust::tuple<int64_t, int64_t>>(1);
+        const auto threshold = -threshold_;
```

Comment:
Why do we need to create an const threshold as minus of threshold_?
Consider add a comment

Reply:
It is due to threshold_ being a class variable, lambda tries to capture the class itself. Adding comment.

Reply:
why minus?

Reply:
So that we don't have to write minus in the code where it is used. The negation operation could also potentially take time. I use negative numbers to count how many times each vertex is cached. Nonnegative numbers to assign local id's in the local cached vertex space.

Reply:
I am a little bit confused about how the negative value is used, how come you can have both positive value to indicate local cached vertex space and negative value to indicate how many times each vertex is accessed? Maybe add comment in _QueryAndIncrement to explain?

Reply:
If a vertex is cached, we don't have to keep its access count anymore. Instead, the hashtable keeps its local id instead of the count.

Thus, if hashtable has a negative value, it means it is not in the cache and keeps the count, else it holds the local id. If the negative value equals -threshold_, then it means we should insert it into the cache during replace operation.

Reply:
Please add docstring for _QueryAndIncrement and explain this logic.